### PR TITLE
Issue #797: support auto-adding timestamp to pubsub messages.

### PR DESCRIPTION
**Pull request by tseaver**
See #797.

### CODE REVIEW

**Comment by tmatsuo** [comment]
This is cool :) Thx!

File: gcloud/pubsub/topic.py [hunk 1]
```diff
@@ -113,6 +122,8 @@ def publish(self, message, **attrs):
         :rtype: str
         :returns: message ID assigned by the server to the published message
         """
+        if self.add_timestamp_to_messages:
+            attrs['timestamp'] = '%sZ' % _NOW().isoformat()
```
Comment (dhermes) [inline gcloud/pubsub/topic.py]: What if `timestamp` is already in `attrs`?

File: gcloud/pubsub/topic.py [hunk 2]
```diff
@@ -39,9 +42,14 @@ class Topic(object):
     :type connection: :class:gcloud.pubsub.connection.Connection
     :param connection: the connection to use.  If not passed,
                         falls back to the default inferred from the
-                        environment.
+
+    :type add_timestamp_to_messages: boolean
+    :param add_timestamp_to_messages: If true, the topic will add a key,
+                        ``timestamp``, to the attributes of each published
```
Comment (dhermes) [inline gcloud/pubsub/topic.py]: Indent doesn't match text. Breaks convention from other docs.

Comment (tseaver) [inline gcloud/pubsub/topic.py]: Length of the argument name made that indentation impractical.  I have verified that this indentation does not break the rendering to HTML (via `tox -e docs && firefox docs/_build/html/pubsub-topic.html`).

File: gcloud/pubsub/topic.py (hunk 1 above)
Comment (tseaver) [inline gcloud/pubsub/topic.py]: We can add code (and extra tests) to handle that case, or just decide that if the user configured the topic with the flag, applying it blindly is the Right Thing (TM).  I had considered both, and decided to go for the simpler case.

File: gcloud/pubsub/topic.py (hunk 2 above)
Comment (dhermes) [inline gcloud/pubsub/topic.py]: Did you consider a shorter argument name? (Convention is pointing out something) It's quite long. How about `auto_timestamp`?

File: gcloud/pubsub/topic.py (hunk 1 above)
Comment (dhermes) [inline gcloud/pubsub/topic.py]: But the same user would have also set the value in the dict so this hypothetical user is likely confused. Silently moving forward is not a great way to help their confusion.

Throw `KeyError` if it's already set?

Comment (tseaver) [inline gcloud/pubsub/topic.py]: I'd rather just leave the explicit value:  see 5ef00ff

File: gcloud/pubsub/topic.py (hunk 2 above)
Comment (tseaver) [inline gcloud/pubsub/topic.py]: 9264828

File: gcloud/pubsub/topic.py (hunk 1 above)
Comment (dhermes) [inline gcloud/pubsub/topic.py]: I'm torn on this.

The hypothetical user goes from being confused that their value was overwritten to confused that their setting `timestamp_messages=True` is not working as expected.

By throwing we say "hey, we don't support doing both of these. you need to pick one"

**Comment by tseaver** [comment]
@dhermes comments addressed, PTAL.

File: gcloud/pubsub/topic.py (hunk 1 above)
Comment (tseaver) [inline gcloud/pubsub/topic.py]: I disagree:  overriding a default at a more local call point is very common (in fact, we do it ourselves with connection / project).  It isn't necessarily a programming error on the user's part:  "consenting adults", etc.

Comment (dhermes) [inline gcloud/pubsub/topic.py]: OK. I'm still torn, but can't find a way to come down on either side, so let's move forward.

**Comment by dhermes** [comment]
LGTM.

**Pull request merged**